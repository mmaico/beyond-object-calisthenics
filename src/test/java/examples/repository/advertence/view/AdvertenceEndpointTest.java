package examples.repository.advertence.view;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import examples.infratest.BaseTest;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.io.IOException;

import static examples.infratest.BDD.*;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.springframework.http.HttpStatus.BAD_REQUEST;
import static org.springframework.http.HttpStatus.CREATED;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;

@ExtendWith({SnapshotExtension.class})
class AdvertenceEndpointTest extends BaseTest {

    private Expect expect;

    @BeforeAll
    public static void beforeAll() throws IOException {
        scenarios.load("advertence-scenario.json");
    }

    @SnapshotName("should create a new advertence")
    @Test
    public void shouldSaveANewAdvertence() throws Exception {
        Given("a new advertence");
            var body = s("new advertence with all valid data");
        When("the api is called");
            var response = mockMvc.perform(post("/advertences")
                    .content(body)
                    .contentType("application/json")).andReturn();
        Then("must return a new advertence created");
            expect.toMatchSnapshot(response.getResponse().getContentAsString());
            assertEquals(CREATED.value(), response.getResponse().getStatus());
    }

    @SnapshotName("should not save because the student not exist")
    @Test
    public void shouldNotSaveBecauseStudentIsInvalid() throws Exception {
        Given("a new advertence");
            var body = s("payload with invalid student");
        When("the api is called");
            var response = mockMvc.perform(post("/advertences")
                .content(body)
                .contentType("application/json")).andReturn();
        Then("must return a bad request");
            expect.toMatchSnapshot(response.getResponse().getContentAsString());
            assertEquals(BAD_REQUEST.value(), response.getResponse().getStatus());
    }
}