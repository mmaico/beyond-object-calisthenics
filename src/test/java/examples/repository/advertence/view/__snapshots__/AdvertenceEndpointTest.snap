should create a new advertence=[
{"_links":{"self":{"href":"/advertences/1"},"applied-by":{"href":"/directors/3"},"received":{"href":"/studants/1"},"class-of":{"href":"/teachers/4"}}}
]



should not save because the student not exist=[
{"uri":"/advertences","errors":{"messages":[{"message":"Student is not found","code":400}]}}
]